### PR TITLE
Crf trainer

### PR DESCRIPTION
**Pull request by aria42**
Refactor CoNLL trainer into a `CRFTrainer` class which can train any model. Have the CoNLL trainer use this class. `CRFTrainer` is tested via the existing roundtrip CoNLL CRF trainer and those tests pass.

### CODE REVIEW

File: src/main/java/org/allenai/ml/sequences/crf/CRFTrainer.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        ExampleObjectiveFn<CRFIndexedExample> objective = new CRFLogLikelihoodObjective<>(weightEncoder);
+        List<CRFIndexedExample> indexedData = labeledData.stream()
+            .map(featureEncoder::indexLabeledExample)
+            .collect(toList());
+        val mrOpts = Parallel.MROpts.withThreads(opts.numThreads);
+        BatchObjectiveFn<CRFIndexedExample> objFn =
+            new BatchObjectiveFn<>(indexedData, objective, weightEncoder.numParameters(), mrOpts);
+        GradientFn regularizer = Regularizer.l2(objFn.dimension(), opts.sigmaSq);
+        val cachedObjFn = new CachingGradientFn(opts.lbfgsHistorySize, objFn.add(regularizer));
+        val quasiNewton = QuasiNewton.lbfgs(opts.lbfgsHistorySize);
```
Comment (chrisc36) [inline src/main/java/org/allenai/ml/sequences/crf/CRFTrainer.java]: Am I reading this correctly to say that `train` only support LBFGS and not gradientDescent? Or would a history of size 0 be equivalent to gradient descent?

Comment (chrisc36) [inline src/main/java/org/allenai/ml/sequences/crf/CRFTrainer.java]: Does this imply `train` only supports LBFGS not vanilla gradient descent? Or would a history of 0 for LBFGS be equal to vanilla gradient descent?

Comment (aria42) [inline src/main/java/org/allenai/ml/sequences/crf/CRFTrainer.java]: 0 means gradient descent. 






On Mon, Jul 27, 2015 at 11:07 AM -0700, "Christopher Clark" <notifications@github.com> wrote:












In src/main/java/org/allenai/ml/sequences/crf/CRFTrainer.java:
> +    {
> +        ensureStartStopPadded(labeledData.stream()
> +                .map(x -> x.stream().map(Pair::getTwo).collect(Collectors.toList()))
> +                .collect(Collectors.toList()),
> +            featureEncoder.stateSpace.startState(),
> +            featureEncoder.stateSpace.stopState());
> +        ExampleObjectiveFn<CRFIndexedExample> objective = new CRFLogLikelihoodObjective<>(weightEncoder);
> +        List<CRFIndexedExample> indexedData = labeledData.stream()
> +            .map(featureEncoder::indexLabeledExample)
> +            .collect(toList());
> +        val mrOpts = Parallel.MROpts.withThreads(opts.numThreads);
> +        BatchObjectiveFn<CRFIndexedExample> objFn =
> +            new BatchObjectiveFn<>(indexedData, objective, weightEncoder.numParameters(), mrOpts);
> +        GradientFn regularizer = Regularizer.l2(objFn.dimension(), opts.sigmaSq);
> +        val cachedObjFn = new CachingGradientFn(opts.lbfgsHistorySize, objFn.add(regularizer));
> +        val quasiNewton = QuasiNewton.lbfgs(opts.lbfgsHistorySize);



Does this imply train only supports LBFGS not vanilla gradient descent? Or would a history of 0 for LBFGS be equal to vanilla gradient descent?

**Comment by chrisc36** [comment]
LGTM, this will make training the model a lot simpler.

**Comment by aria42** [comment]
Hey @chrisc36, I think f466bdb fixes #13 , mind taking a look

**Comment by chrisc36** [comment]
That looks fine to me

**Pull request merged**